source.go

```diff
@@ -303,12 +338,28 @@ func WithCreateDestPath() llb.CopyOption {
 	})
 }
 
-func SourceIsDir(src Source) bool {
+func (s *Source) IsDir(srcState llb.State, sOpt SourceOpts) bool {
+	return SourceIsDir(*s, srcState, sOpt)
+}
+
+func SourceIsDir(src Source, srcState llb.State, sOpt SourceOpts) bool {
 	switch {
+	case src.Context != nil:
+		if !isRoot(src.Path) {
+			srcFS := sOpt.GetFS(srcState)
+			entries, err := srcFS.ReadDir("/")
```

Comment:
The number of entries is not what matters here, only if `src.Path` is pointing to a file or a dir.

Reply:
the state here already represents the merged state so the `src.Path` is not longer valid and we wanted to check the `/` root

Reply:
In that case its not really a valid test because `Path` can point to a directory with a single file in it.
So we need to check the origin state before its fully filtered down to a path.